### PR TITLE
[FXML-4501] Lower arith.andi, arith.ori, arith.shli, arith.shrsi, arith.shrui, arith.xori to emitc

### PR DESCRIPTION
**Pull request by cferry-AMD**

### CODE REVIEW

File: mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir [hunk 1]
```diff
@@ -48,6 +48,16 @@ func.func @arith_integer_ops(%arg0: i32, %arg1: i32) {
   %3 = arith.divsi %arg0, %arg1 : i32
   // CHECK: emitc.rem %arg0, %arg1 : (i32, i32) -> i32
   %4 = arith.remsi %arg0, %arg1 : i32
+  // CHECK: emitc.bitwise_and %arg0, %arg1 : (i32, i32) -> i32
```
Comment (mgehre-amd) [inline mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir:51]: I think we should always cast to unsigned types for bitwise operations (except `shrsi`, which is explicitly signed) as
the C11 standard, section 6.5, paragraph 4 states:

```
Some operators (the unary operator ~ , and the binary operators <<, >>, &, ^, and |, collectively described as bitwise operators) shall have operands that have integral type. These operators return values that depend on the internal representations of integers, and thus have implementation-defined and undefined aspects for signed types.
```

File: mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir [hunk 2]
```diff
@@ -48,6 +48,16 @@ func.func @arith_integer_ops(%arg0: i32, %arg1: i32) {
...
+  %5 = arith.andi %arg0, %arg1 : i32
+  // CHECK: emitc.bitwise_or %arg0, %arg1 : (i32, i32) -> i32
+  %6 = arith.ori %arg0, %arg1 : i32
+  // CHECK: emitc.bitwise_xor %arg0, %arg1 : (i32, i32) -> i32
+  %7 = arith.xori %arg0, %arg1 : i32
+  // CHECK: emitc.bitwise_left_shift %arg0, %arg1 : (i32, i32) -> i32
+  %8 = arith.shli %arg0, %arg1 : i32
+  // CHECK: emitc.bitwise_right_shift %arg0, %arg1 : (i32, i32) -> i32
+  %9 = arith.shrsi %arg0, %arg1 : i32
```
Comment (mgehre-amd) [inline mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir:60]: While you are at it, maybe also support `shrui`?

File: mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir (hunk 1 above)
Comment (cferry-AMD) [reply mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir:51]: In this case I'd be inclined to wait for us to bump LLVM before carrying on working on this, as the conversion to unsigned seems to be needed in numerous operations now.

File: mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir (hunk 2 above)
Comment (cferry-AMD) [reply mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir:60]: `shrui` requires a conversion to unsigned. I didn't include operations that needed this (and, well, turns out I missed on some).

**Comment by mgehre-amd** [comment]
Alternatively, checking the `nuw/nsw` flags on arith.shli might guarantee that the shift amount is valid. I haven't looked to much which of them is applicable here.

**Comment by mgehre-amd** [comment]
> Alternatively, checking the `nuw/nsw` flags on arith.shli might guarantee that the shift amount is valid. I haven't looked to much which of them is applicable here.

Ah, those flags don't constraint the shift amount. A better description of them is in the LLVM spec: https://llvm.org/docs/LangRef.html#shl-instruction

**Comment by cferry-AMD** [comment]
Static checking is only possible when the amount of the right shift is statically known, which practically means a constant. So, in the general case we'll need to introduce runtime checks.

However, I'm not sure of how to deal with anything that relies on poison values. There's no poison semantics in C/C++, and I have too little understanding of the differences between poison and undefined behavior to actually know what to do.

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp [hunk 3]
```diff
@@ -443,6 +444,144 @@ class IntegerOpConversion final : public OpConversionPattern<ArithOp> {
...
+
+    // There is no unsigned i1 type, bitwise ops can be performed directly
+    // on booleans.
+    if (type.isInteger(1)) {
+      rewriter.replaceOpWithNewOp<EmitCOp>(op, type, adaptor.getLhs(),
+                                           adaptor.getRhs());
+      return success();
+    }
+
+    Type arithmeticType = adaptIntegralTypeSignedness(type, true);
```
Comment (mgehre-amd) [inline mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:471]:
```suggestion
    Type arithmeticType = adaptIntegralTypeSignedness(type, /*isUnsigned=*/true);
```

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp [hunk 4]
```diff
@@ -443,6 +444,144 @@ class IntegerOpConversion final : public OpConversionPattern<ArithOp> {
...
+                  ConversionPatternRewriter &rewriter) const override {
+
+    Type type = this->getTypeConverter()->convertType(op.getType());
+    if (!isa_and_nonnull<IntegerType, emitc::SignedSizeTType, emitc::SizeTType>(
+            type)) {
+      return rewriter.notifyMatchFailure(
+          op, "expected integer or size_t/ssize_t type");
+    }
+
+    // There is no unsigned i1 type, bitwise ops can be performed directly
```
Comment (mgehre-amd) [inline mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:463]: I don't understand this comment. The type `ui1` seems to exist.

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp [hunk 5]
```diff
@@ -443,6 +444,144 @@ class IntegerOpConversion final : public OpConversionPattern<ArithOp> {
...
+      return rewriter.notifyMatchFailure(op, "i1 type is not implemented");
+    }
+
+    Type arithmeticType = adaptIntegralTypeSignedness(type, isUnsignedOp);
+
+    Value lhs = adaptValueType(adaptor.getLhs(), rewriter, arithmeticType);
+    Value rhs = adaptValueType(adaptor.getRhs(), rewriter, arithmeticType);
+
+    // Add a runtime check for overflow
+    // This is an abuse of the size_t type since we're potentially using values
```
Comment (mgehre-amd) [inline mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:512]: The comment is not clear to me. What is it trying to say?

**Comment by cferry-AMD** [comment]
In the runtime check, I resorted to using `sizeof` for index types instead of rejecting them. I expect that the compiler does constant folding and strength reduction when the shift amount is a constant, to avoid introducing a shortcut.

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp [hunk 6]
```diff
@@ -443,6 +444,144 @@ class IntegerOpConversion final : public OpConversionPattern<ArithOp> {
...
+      width = rewriter.create<emitc::ConstantOp>(
+          op.getLoc(), sizeTType,
+          rewriter.getIntegerAttr(sizeTType, type.getIntOrFloatBitWidth()));
+    }
+
+    Value oobTest;
+    Value excessCheck = rewriter.create<emitc::CmpOp>(
+        op.getLoc(), rewriter.getI1Type(), emitc::CmpPredicate::lt, rhs, width);
+    oobTest = excessCheck;
+    if (!isUnsignedOp) {
```
Comment (mgehre-amd) [inline mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:536]: The second operand is always interpreted as unsigned, even for https://mlir.llvm.org/docs/Dialects/ArithOps/#arithshrsi-arithshrsiop: `The first operand is interpreted as signed, and the second operand is interpreter as unsigned.`.

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp (hunk 5 above)
Comment (cferry-AMD) [reply mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:512]: `ssize_t` can only represent whatever `size_t` can, plus -1. However, it almost always turns out it is just turned into the longest built-in signed int, and so it supports other negative values. If the source `index` is below -1 (which is perfectly legal), turning it into `ssize_t` will work but the value is outside the range of officially supported values for that type.

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp [hunk 7]
```diff
@@ -443,6 +444,144 @@ class IntegerOpConversion final : public OpConversionPattern<ArithOp> {
...
+          (isa<IntegerType>(sizeTType) ? rewriter.getIntegerAttr(sizeTType, 0)
+                                       : rewriter.getIndexAttr(0)));
+      Value defaultCheck =
+          rewriter.create<emitc::CmpOp>(op.getLoc(), rewriter.getI1Type(),
+                                        emitc::CmpPredicate::ge, rhs, zero);
+      oobTest = rewriter.create<emitc::LogicalAndOp>(
+          op.getLoc(), rewriter.getI1Type(), excessCheck, defaultCheck);
+    }
+
+    Value poison = rewriter.create<emitc::ConstantOp>(
```
Comment (mgehre-amd) [inline mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:548]:
```suggestion
    // Any concrete value is a valid refinement of poison.
    Value poison = rewriter.create<emitc::ConstantOp>(
```

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp [hunk 8]
```diff
@@ -443,6 +444,144 @@ class IntegerOpConversion final : public OpConversionPattern<ArithOp> {
...
+        op.getLoc(), arithmeticType,
+        (isa<IntegerType>(arithmeticType)
+             ? rewriter.getIntegerAttr(arithmeticType, 0)
+             : rewriter.getIndexAttr(0)));
+
+    emitc::ExpressionOp ternary = rewriter.create<emitc::ExpressionOp>(
```
Comment (mgehre-amd) [inline mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:554]: Nice use of `emitc::ExpressionOp`. What does the generated C++ code look like?

File: mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir [hunk 9]
```diff
@@ -92,6 +92,166 @@ func.func @arith_index(%arg0: i32, %arg1: i32) {
...
+  %2 = arith.shrui %arg0, %arg1 : i32
+
+  // CHECK-DAG: %[[SSizeConstant:[^ ]*]] = "emitc.constant"{{.*}}value = 32{{.*}}i32
+  // CHECK-DAG: %[[SCmpNoExcess:[^ ]*]] = emitc.cmp lt, %[[ARG1]], %[[SSizeConstant]] : (i32, i32) -> i1
+  // CHECK-DAG: %[[SZeroConstant:[^ ]*]] = "emitc.constant"{{.*}}value = 0{{.*}}i32
+  // CHECK-DAG: %[[SCmpPositive:[^ ]*]] = emitc.cmp ge, %[[ARG1]], %[[SZeroConstant]] : (i32, i32) -> i1
+  // CHECK-DAG: %[[SCmpAnd:[^ ]*]] = emitc.logical_and %[[SCmpNoExcess]], %[[SCmpPositive]] : i1, i1
+  // CHECK-DAG: %[[SZero:[^ ]*]] = "emitc.constant"{{.*}}value = 0{{.*}}i32
+  // CHECK: %[[SShiftRes:[^ ]*]] = emitc.expression : i32 {
+  // CHECK: %[[SHRSI:[^ ]*]] = emitc.bitwise_right_shift %[[ARG0]], %[[ARG1]] : (i32, i32) -> i32
```
Comment (mgehre-amd) [inline mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir:164]: I would expect something like
```suggestion
  // CHECK: %[[SHRSI:[^ ]*]] = emitc.bitwise_right_shift %[[ARG0]], %[[ARG1]] : (si32, ui32) -> si32
```

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp (hunk 5 above)
Comment (mgehre-amd) [reply mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:512]: This seems to be applicable wherever we decide to convert index to ssize_t, i.e. mainly in the type converter. Why is that comment in this place specifically?

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp (hunk 8 above)
Comment (cferry-AMD) [reply mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:554]: The shift operation (not just its result) must be in the same expression as the ternary op to benefit from the conditional evaluation and avoid UB, hence this use of `ExpressionOp`.

```C++
void shift() {
  // Constants and their signed casts
  size_t v1 = 160;
  size_t v2 = 64;
  ssize_t v3 = (ssize_t) v1;
  ssize_t v4 = (ssize_t) v2;

  // Here goes the right shift op
  // Predicate : (shift amount < 8*sizeof(ssize_t)) && (shift amount >= 0)
  ssize_t v5 = 8;
  ssize_t v6 = sizeof(v5);
  ssize_t v7 = v5 * v6;
  bool v8 = v4 < v7;
  ssize_t v9 = 0;
  bool v10 = v4 >= v9;
  bool v11 = v8 && v10;

  ssize_t v12 = 0; // Poison value
  // Actual shift op.
  size_t v13 = (size_t) v11 ? v3 >> v4 : v12;
  return;
}
```

Comment (mgehre-amd) [reply mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:554]: Looks good! On the last line `size_t v13 = (size_t) v11 ? v3 >> v4 : v12;`, why is there a cast? `v11` is already bool and that is the type the conditional parts needs to have.

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp (hunk 4 above)
Comment (cferry-AMD) [reply mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:463]: I had support for ui1s until I read somewhere in the codebase that this type was illegal, and I then removed that code. I can't find that place anymore, plus ui1 indeed seems to exist. 

Nevertheless, we don't need to perform the conversion here anyway for i1s, so the comment can go and the behavior stay.

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp (hunk 5 above)
Comment (cferry-AMD) [reply mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:512]: We don't convert index to `ssize_t` but to `size_t` in the type converter. Now you're right, this comment is general.

File: mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir (hunk 9 above)
Comment (cferry-AMD) [reply mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir]: We don't turn signless `i32`s into `si32`s yet, and I'm expecting this change to have a broader impact than just this op, so I suggest to do that in a separate PR ("turn signless integers into signed, remove support for signless integers in EmitC" or something along those lines).

Comment (mgehre-amd) [reply mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir]: Of course, you are right. Sorry for the noise.

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp (hunk 8 above)
Comment (cferry-AMD) [reply mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:541]: Per [this page](https://en.cppreference.com/w/cpp/language/operator_precedence), the cast is on `v11` indeed, where in the MLIR it is on the result of the expression. That looks a bug in the C++ emission phase.

The EmitC MLIR
```mlir
    %8 = emitc.expression : !emitc.ssize_t {
      %10 = emitc.bitwise_right_shift %2, %1 : (!emitc.ssize_t, !emitc.size_t) -> !emitc.ssize_t
      %11 = emitc.conditional %6, %10, %7 : !emitc.ssize_t
      emitc.yield %11 : !emitc.ssize_t
    }
    %9 = emitc.cast %8 : !emitc.ssize_t to !emitc.size_t
```

is turned into this C++
```c++
size_t v9 = (size_t) v7 ? v3 >> v2 : v8;
```

so it looks like we're missing some parentheses.

Comment (cferry-AMD) [reply mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:541]: [Raised an issue](https://github.com/llvm/llvm-project/issues/93470) on upstream LLVM.

**Review by josel-amd — APPROVED** [review]
LGTM

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp [hunk 10]
```diff
@@ -443,6 +444,131 @@ class IntegerOpConversion final : public OpConversionPattern<ArithOp> {
...
+class BitwiseOpConversion : public OpConversionPattern<ArithOp> {
+public:
+  using OpConversionPattern<ArithOp>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(ArithOp op, typename ArithOp::Adaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+
+    Type type = this->getTypeConverter()->convertType(op.getType());
+    if (!isa_and_nonnull<IntegerType, emitc::SignedSizeTType, emitc::SizeTType>(
```
[the remark this answers is not in the document]
Comment (cferry-AMD) [reply mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:457]: Not really. I just did the same as in the rest of this pass :)

File: mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir [hunk 11]
```diff
@@ -92,6 +92,161 @@ func.func @arith_index(%arg0: i32, %arg1: i32) {
...
+  // CHECK-DAG: %[[C1:[^ ]*]] = emitc.cast %[[ARG0]] : i32 to ui32
+  // CHECK-DAG: %[[C2:[^ ]*]] = emitc.cast %[[ARG1]] : i32 to ui32
+  // CHECK-DAG: %[[SizeConstant:[^ ]*]] = "emitc.constant"{{.*}}value = 32
+  // CHECK-DAG: %[[CmpNoExcess:[^ ]*]] = emitc.cmp lt, %[[C2]], %[[SizeConstant]] : (ui32, ui32) -> i1
+  // CHECK-DAG: %[[Zero:[^ ]*]] = "emitc.constant"{{.*}}value = 0
+  // CHECK: %[[ShiftRes:[^ ]*]] = emitc.expression : ui32 {
+  // CHECK: %[[SHL:[^ ]*]] = emitc.bitwise_left_shift %[[C1]], %[[C2]] : (ui32, ui32) -> ui32
+  // CHECK: %[[Ternary:[^ ]*]] = emitc.conditional %[[CmpNoExcess]], %[[SHL]], %[[Zero]] : ui32
+  // CHECK: emitc.yield %[[Ternary]] : ui32
+  // CHECK: }
```
Comment (TinaAMD) [inline mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir:133]: Can you drop the braces (opening and closing)?

File: mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir [hunk 12]
```diff
@@ -92,6 +92,161 @@ func.func @arith_index(%arg0: i32, %arg1: i32) {
...
+  // CHECK: emitc.cast %[[SShiftRes]] : !emitc.ssize_t to !emitc.size_t
+  %3 = arith.shrsi %arg0, %arg1 : index
+
+  return
+}
+
+// -----
+
+// CHECK-LABEL: arith_bitwise_bool
+func.func @arith_bitwise_bool(%arg0: i1, %arg1: i1) {
```
Comment (TinaAMD) [inline mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir:238]: Can you move this up to the other bitwise tests?

**Review by TinaAMD — APPROVED** [review]
LGTM

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp [hunk 13]
```diff
@@ -443,6 +444,131 @@ class IntegerOpConversion final : public OpConversionPattern<ArithOp> {
...
+            type)) {
+      return rewriter.notifyMatchFailure(
+          op, "expected integer or size_t/ssize_t type");
```
[the remark this answers is not in the document]
Comment (TinaAMD) [reply mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:461]: I would put it in the error message, then we don't need an additional comment 😅

**Pull request merged**